Next.mk

```diff
@@ -37,32 +39,110 @@ build: $(current_binary) ## Build the func-e binary
-$(build_dir)/func-e_darwin_%/func-e: override goos=darwin
-$(build_dir)/func-e_linux_%/func-e: override goos=linux
-$(build_dir)/func-e_%_amd64/func-e: override goarch=amd64
-$(build_dir)/func-e_%_arm64/func-e: override goarch=arm64
-$(build_dir)/func-e_%/func-e:
-	$(call go-build, $@)
+non_windows_platforms := darwin_amd64 darwin_arm64 linux_amd64 linux_arm64
+# TODO: arm64 on Windows https://github.com/envoyproxy/envoy/issues/17572
+windows_platforms := windows_amd64
 
-$(windows_binary): override goos=windows
-$(windows_binary): override goarch=amd64
-$(windows_binary):
-	$(call go-build, $@)
+$(build_dir)/func-e_%/func-e: main.go
+	$(call go-build, $@, $^)
 
+$(dist_dir)/func-e_$(VERSION)_%.tar.gz: $(build_dir)/func-e_%/func-e
+	@printf "$(ansi_format_dark)" tar.gz "tarring $@"
+	@mkdir -p $(dir $@)
+	@tar --strip-components 2 -cpzf $@ $<
+	@printf "$(ansi_format_bright)" tar.gz "ok"
+
+$(build_dir)/func-e_%/func-e.exe: main.go
+	$(call go-build, $@, $^)
+
+$(dist_dir)/func-e_$(VERSION)_%.zip: $(build_dir)/func-e_%/func-e.exe.signed
+	@printf "$(ansi_format_dark)" zip "zipping $@"
+	@mkdir -p $(dir $@)
+ifeq ($(goos),windows)  # Windows 10 etc use 7zip
+	@# '-bso0 -bsp0' makes it quiet, except errors
+	@# Wildcards in 7zip will skip the directory. We single-quote to ensure they aren't interpreted by the shell.
+	@7z -bso0 -bsp0 a $@ './$(dir $<)/*.exe'
+else  # Otherwise, assume zip is available
+	@zip -qj $@ $(<:.signed=)
+endif
+	@printf "$(ansi_format_bright)" zip "ok"
+
+# msi-arch is a macro so we can detect it based on the file naming convention
+msi-arch = $(if $(findstring amd64,$1),x64,arm64)
+# Default to a dummy version, which is always lower than a real release
+msi_version=$(VERSION:dev=0.0.1)
+
+# This builds the Windows installer (MSI) using platform-dependent WIX commands.
+$(dist_dir)/func-e_$(VERSION)_%.msi: $(build_dir)/func-e_%/func-e.exe.signed
+	@printf "$(ansi_format_dark)" msi "building $@"
+	@mkdir -p $(dir $@)
+ifeq ($(goos),windows)  # Windows 10 etc use https://wixtoolset.org
+	@candle -nologo -arch $(call msi-arch,$@) -dVersion=$(msi_version) -dBin=$(<:.signed=) packaging/msi/func-e.wxs
+	@light -nologo func-e.wixobj -o $@ -spdb
+	@rm func-e.wixobj
+else  # use https://wiki.gnome.org/msitools
+	@wixl -a $(call msi-arch,$@) -D Version=$(msi_version) -D Bin=$(<:.signed=) -o $@ packaging/msi/func-e.wxs
+endif
+	$(call codesign, $@)
+	@printf "$(ansi_format_bright)" msi "ok"
+
+# Archives are tar.gz, except in the case of Windows, which uses zip.
+archives  := $(non_windows_platforms:%=$(dist_dir)/func-e_$(VERSION)_%.tar.gz) $(windows_platforms:%=$(dist_dir)/func-e_$(VERSION)_%.zip)
+packages  := $(windows_platforms:%=$(dist_dir)/func-e_$(VERSION)_%.msi)
+checksums := $(dist_dir)/func-e_$(VERSION)_checksums.txt
+
+# Darwin doesn't have sha256sum. See https://github.com/actions/virtual-environments/issues/90
+sha256sum := $(if $(findstring darwin,$(goos)),shasum -a 256,sha256sum)
```

Comment:
this really should be `(if (eq`, as it will be a keyword equal.  However, this works, and I'm not sure the function if any for inline equals https://www.gnu.org/software/make/manual/html_node/Text-Functions.html